engines/freescape/loaders/8bitBinaryLoader.cpp

```diff
@@ -0,0 +1,590 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// Based on Phantasma code by Thomas Harte (2013)
+
+#include "common/file.h"
+#include "image/bmp.h"
+
+#include "freescape/freescape.h"
+#include "freescape/language/8bitDetokeniser.h"
+#include "freescape/neo.h"
+#include "freescape/objects/global.h"
+#include "freescape/objects/sensor.h"
+
+namespace Freescape {
+
+uint16 FreescapeEngine::readField(Common::SeekableReadStream *file, int bits) {
+	uint16 value;
+	assert(bits == 8 || bits == 16);
+	if (isAmiga() || isAtariST()) {
+		if (bits == 16) {
+			uint16 lo = file->readUint16BE();
+			assert(lo < 256);
+			uint16 hi = file->readUint16BE();
+			assert(hi < 256);
+			value = 256 * hi + lo;
+			value = 2 * value; // Unclear why, but this reads a pointer
+		} else {
+			assert(bits == 8);
+			value = file->readUint16BE();
+			if (value >= 256) {
+				warning("failed to read byte with value 0x%x", value);
+				value = value & 0xff;
+			}
+		}
+	} else {
+		if (bits == 8)
+			value = file->readByte();
+		else
+			value = file->readUint16LE();
+	}
+
+	return value;
+}
+
+Common::Array<uint8> FreescapeEngine::readArray(Common::SeekableReadStream *file, int size) {
+	byte *data = (byte *)malloc(size);
+	for (int i = 0; i < size; i++) {
+		data[i] = readField(file, 8);
+	}
+	Common::Array<uint8> array(data, size);
+	return array;
+}
+
+Object *FreescapeEngine::load8bitObject(Common::SeekableReadStream *file) {
+
+	byte rawFlagsAndType = readField(file, 8);
+	debugC(1, kFreescapeDebugParser, "Raw object data flags and type: %d", rawFlagsAndType);
+	Object::Type objectType = (Object::Type)(rawFlagsAndType & 0x1F);
+
+	Math::Vector3d position, v;
+
+	position.x() = readField(file, 8);
+	position.y() = readField(file, 8);
+	position.z() = readField(file, 8);
+
+	v.x() = readField(file, 8);
+	v.y() = readField(file, 8);
+	v.z() = readField(file, 8);
+
+	// object ID
+	uint16 objectID = readField(file, 8);
+	// size of object on disk; we've accounted for 8 bytes
+	// already so we can subtract that to get the remaining
+	// length beyond here
+	uint8 byteSizeOfObject = readField(file, 8);
+	debugC(1, kFreescapeDebugParser, "Raw object %d ; type %d ; size %d", objectID, (int)objectType, byteSizeOfObject);
+	if (byteSizeOfObject < 9) {
+		error("Not enough bytes %d to read object %d with type %d", byteSizeOfObject, objectID, objectType);
+	}
+
+	assert(byteSizeOfObject >= 9);
+	byteSizeOfObject = byteSizeOfObject - 9;
+	if (objectID == 255 && objectType == Object::Type::Entrance) {
+		debugC(1, kFreescapeDebugParser, "Found the room structure (objectID: 255 with size %d)", byteSizeOfObject + 6);
+		byte *structureData = (byte *)malloc(byteSizeOfObject + 6);
```

Comment:
I think this leaks memory.

Reply:
Fixed. Note that the `while(..) { array->push_back(..)` is only used with a small number of iterations so it shouldn't be too slow.